activerecord/lib/active_record/enum.rb

```diff
@@ -128,13 +159,19 @@ def cast(value)
 
       def deserialize(value)
         return if value.nil?
+        return value if string_enum? && mapping.keys.include?(value)
         mapping.key(subtype.deserialize(value))
       end
 
       def serialize(value)
+        return value if string_enum? && mapping.keys.include?(value)
         mapping.fetch(value, value)
       end
 
+      def string_enum?
+        mapping.first&.map(&:class) == [String, Integer] && @subtype.type == :string
+      end
```

Comment:
Is there no way to make this backwardscompatible without moving these checks to runtime? I don't find this new code an improvement.

I mean, should we automatically detect that a column changed type between boots? Don't think we do that for other columns that change type for instance.

Reply:
I can suggest two ways of reducing a number of checks:

1. I can cache it with `@_string_enum ||=`
2. I can add a special option that will turn this behaviour on as I suggested above (https://github.com/rails/rails/pull/32241#discussion_r206305526). But, as I mentioned above, It will kill the magic.

Reply:
> should we automatically detect that a column changed type between boots?

I'm confused by this @kaspth - did you mean to write "between requests?" Accessing `@subtype.type` won't reload the schema from the database, it's a property of the type object.

Reply:
Here's the implementation I had in mind:

```patch
diff --git a/activerecord/lib/active_record/enum.rb b/activerecord/lib/active_record/enum.rb
index 23ecb24542..7ab8f3177e 100644
--- a/activerecord/lib/active_record/enum.rb
+++ b/activerecord/lib/active_record/enum.rb
@@ -110,7 +110,7 @@ class EnumType < Type::Value # :nodoc:

       def initialize(name, mapping, subtype)
         @name = name
-        @mapping = mapping
+        @mapping = resolve_mapping(mapping, subtype)
         @subtype = subtype
       end

@@ -143,6 +143,16 @@ def assert_valid_value(value)

       private
         attr_reader :name, :mapping, :subtype
+
+        def resolve_mapping(mapping, subtype)
+          if mapping.respond_to?(:each_pair)
+            mapping
+          elsif subtype.type == :string
+            mapping.index_by(&:itself)
+          else
+            mapping.each_with_index.to_h
+          end.with_indifferent_access
+        end
     end

     def enum(definitions)
@@ -164,7 +174,7 @@ def enum(definitions)

         attr = attribute_alias?(name) ? attribute_alias(name) : name
         decorate_attribute_type(attr, :enum) do |subtype|
-          EnumType.new(attr, enum_values, subtype)
+          EnumType.new(attr, values, subtype)
         end

         _enum_methods_module.module_eval do
```

The call to `decorate_attribute_type` ensures that this code won't run until the schema is loaded, and we still only do the work once - we just do it later.

Reply:
@eugeneius I was thinking about what would happen when new code is loaded after a deploy. So "between deploy boots" would probably have been clearer.

Dig the code suggestion 👍 (we'd need to require a core_ext for `index_with` too).

Reply:
Your realization (https://github.com/rails/rails/pull/32241/commits/5bbeac98aa3ba3a616cebe7fd566d36c4fd7d230#r206364464) is fine. I've tried to adopt it to my code. But I found two problems:

1. If the enum field was defined using an array (`enum cover: %i[hardcover paperback]), the code in `resolve_mapping` will break the behaviour of autogenerated bang-methods, question-mark-methods and scope-methods, cause they are defined when the actual type of the field is not yet known and those they will always try to use integers instead of string values. The easiest way to provide a proper mapping to keep this methods working is to keep `integer-to-string` mapping logic inside of `#serialize`/`#deserialize` methods.

I mean this code below:


```ruby
# activerecord/lib/active_record/enum.rb  

...

# def active?() status == "active" end
klass.send(:detect_enum_conflict!, name, "#{value_method_name}?")
define_method("#{value_method_name}?") { self[attr] == label }

# def active!() update!(status: 0) end
klass.send(:detect_enum_conflict!, name, "#{value_method_name}!")
define_method("#{value_method_name}!") { update!(attr => value) }

# scope :active, -> { where(status: 0) }
klass.send(:detect_enum_conflict!, name, value_method_name, true)
klass.scope value_method_name, -> { where(attr => value) }

...

```

2. Both: lables and values for the hash-definition (`enum cover: { hardcover: :hardcover, paperback: :paperback }`) of enum field can be defined either by a String or by a Symbol. Some additional logic is needed to convert symbols to strings (`HashWithIndifferentAccess` is not enought, cause values should also be mapped to strings, not only keys).


Considering both problematic cases described above we would need some additional code which will also lead to a larger conditional nesting and increase of complexity.


But your example pointed me to a variable `values`, which holds the original definition of the enum field which is extremely helpful (https://github.com/rails/rails/pull/32241/commits/89820b00c5b35d7411b030b082d21f937f8eeca3#diff-abb244a8c26afeef0165452650430244R185) for determining if the `enum` was defined with an Array or with a Hash.

Thank you!